### ISSUE TITLE
clicol, cu and USB adapter

### ISSUE THREAD

**Issue by Raizo62** [realvitya/clicol#14]
Hi

I wish to use clicol with serial console

I have installed the command cu (from the stable Debian package)

When i do : `clicol-cmd cu -l /dev/ttyS0`
I see the console output with (cisco) colors. It is ok.

But, when i tested with USB adapter for serial console, i used the command :
`clicol-cmd cu -l /dev/ttyUSB0`
Strangely, I see the good console output but without color.

Do you know why ? a bug in clicol ? must I use another command than cu (which one?) ? or another command of clicol (how?) ?

**Comment by realvitya** [comment]
Hey,
Did you try minicom?
It seems to be a buffering issue. I have no experience with serial yet.

Regards
Viktor

**Comment by Raizo62** [comment]
minicom is the first that i have tried. But it seems more slow to response when it is start with cli-cmd :-(

**Comment by realvitya** [comment]
If it’s slow then please set the maxwait timer to a very low setting like
0.3
You can do it in the clicol.cfg file. Play with that variable!

Regards,
Viktor

**Comment by Raizo62** [comment]
Hi

Minicom has the same problem that cu : output is colorize with serial, but not with usb adapter.

Can you say me where do you read the data, in the code ? i can perhaps debug and see how to solve the problem.....

**Comment by realvitya** [comment]
Try the break key then shift+d then repeat. This enables the most verbose
debug. If you can capture the output for me I'd look on it.

Viktor

**Comment by Raizo62** [comment]
Hi

Thanks you.

I didn't success to do the break key and shift+d.
Instead, i have modified the value debug=1 in the .clicol/clicol.cfg

Here the 2 files are the copy-paste.
In the 2 files, the debug infos were in "brown/orange" color.

[debugUSB.txt](https://github.com/realvitya/clicol/files/4329600/debugUSB.txt)
[debugSerial.txt](https://github.com/realvitya/clicol/files/4329601/debugSerial.txt)

If you really need the break method, i will investigate more

**Comment by realvitya** [comment]
It looks the same as I thought. It seems like usb does not do any kind of
buffering and send the output char by char which makes it hard to put
together into a line. I guess in that case if you set maxwait to a higher
number than the time between 2 character appearing, it would be colored.
But it may be slow like hell.
So basically it’s rather the input is coming other way. I’d try stdbuf
command with line buffering to see if that’s working.

Thanks!

**Comment by Raizo62** [comment]
Perhaps that the buffer will correct also the bug where somes lines (but few) are not colored (in normal use)

**Comment by realvitya** [comment]
If you see lines like that, can you please provide me info about:
Command that causes the output, the rtr time to the device, device type
I saw those only when using serial or console in console. Problem is the
way the device send the output, I have to estimate when the output ends so
I can colorize and display the output to user. Until that I do buffering.
So there is a compromise between correct output and response time. That's
what that maxwait seeing is about. I was thinking about measure the rtr and
set this dynamically. Not yet implemented.

Thanks!

**Comment by Raizo62** [comment]
> If you see lines like that, can you please provide me info about:

Sorry, what are you talking about ? "somes lines (but few) are not colored" or the problem with usb ?

"somes lines (but few) are not colored" was reproduce on Cisco Router in GNS3/dynamips. As you can see, i believed that i couldn't reproduce it.
![2020-03-16-130350_1280x1024_scrot](https://user-images.githubusercontent.com/7449584/76757070-3a9f8480-6787-11ea-909f-69c971b24c14.png)

Where do i find the rtr time ?

**Comment by realvitya** [comment]
rtr is like you ping the device. I could reproduce the problem in GNS3 with slow link. And it seems indeed a bug in code. I need to fire up my linux dev machine to debug further. Give me some time for that!

**Comment by Raizo62** [comment]
Thanks :-)

**Comment by realvitya** [comment]
@Raizo62 may I ask you to test with this latest commit? By me, problem went away.

**Comment by Raizo62** [comment]
sorry, i have tested (on true device) : i see no difference (few lines are not colored) :-(
I have change nothing in the config. Was I supposed to change anything ?

**Comment by realvitya** [comment]
No, except for pulling the latest commit from master branch...
What is your maxwait value? May I ask you to set it to 2.0 or more?

**Comment by Raizo62** [comment]
maxwait is the default value.

**Comment by realvitya** [comment]
Did you try with higher value? In the clicol.cfg file.

**Comment by Raizo62** [comment]
with 3 routers on GNS3.
maxwait=4.0

![2020-03-19-170840_1920x1080_scrot](https://user-images.githubusercontent.com/7449584/77089458-2321ff00-6a06-11ea-8ecf-6735951bbdb2.png)
![2020-03-19-171933_1920x1080_scrot](https://user-images.githubusercontent.com/7449584/77089525-38972900-6a06-11ea-8236-02eaf2ec1681.png)

But it not easy to reproduce.
Except the first interface of the router with the command "show run" : "always" un-colored.

**Comment by realvitya** [comment]
Last question, you used the latest version right?

**Comment by Raizo62** [comment]
yes, i believe... my last commit is "trying fixing timeout related bug"

**Comment by Raizo62** [comment]
Here, the debug mode of "fastethernet0/0" without color.
![display](https://user-images.githubusercontent.com/7449584/77157247-451c8f80-6aa1-11ea-8444-574c77ff9d28.png)

[debug log.txt](https://github.com/realvitya/clicol/files/4359104/fastethernet0_0_without_color.txt)

But, my big problem is more the USB : few PC have still an serial port. We must use the USB adapter.

**Comment by realvitya** [comment]
So for USB. I checked again the printout logic and most probably you end up
in this code:
line329:

>                 elif charbuffer == inputtext:  # interactive
>                     charbuffer = ""
>                     # colorize only short stuff (up key,ping)
>                     return colorize(bufout, ["prompt",
> "ping"]).encode('utf-8')
>                 else:  # need to collect more output
>                     return b""

If you comment out the elif part, you may end up in buffering for single
character input and you may even see colors. The problem with that would be
that when you type, your characters will appear after maxwait time which
hurts the user experience. I started thinking on something like we need to
watch if the device is in output mode. Like if we see incoming bits, we
start buffering but if we did not see input since a while, we show output
immediately like in the above code snippet. (and what you experience)

May I ask you for the sake of test, would you comment out the elif part and
test with the USB? Don't worry about that what you type is not appearing
just enter the show command. I'm curious if the output would be colorized
after that. Thanks!

Raizo62 <notifications@github.com> ezt írta (időpont: 2020. márc. 20., P,
11:56):

**Comment by Raizo62** [comment]
I have tested with the usb : the output is colored (and what i type is not appearing immediately)

**Comment by realvitya** [comment]
Ok, thanks. Give me some time to figure out!

**Comment by realvitya** [comment]
@Raizo62 may I ask you to checkout usbtest branch and test again with USB?
If first couple of lines are not colored, please decrease maxwait to lower value!
Sorry, I can't test this, I don't have real device to test here. If something is weird, please show me debug output! txt is enough...

**Comment by Raizo62** [comment]
- With maxwait=default
![maxwait=default](https://user-images.githubusercontent.com/7449584/77246120-5b9f2400-6c24-11ea-81fe-b5f28a939261.png)
[error_maxwait=default.txt](https://github.com/realvitya/clicol/files/4364916/error_maxwait.default.txt)
[error_copy_paste_maxwait=default.txt](https://github.com/realvitya/clicol/files/4364917/error_copy_paste_maxwait.default.txt)

- With maxwait=0.0
![maxwait=0](https://user-images.githubusercontent.com/7449584/77246124-5e017e00-6c24-11ea-8628-8c2aedb9626b.png)

**Comment by realvitya** [comment]
@Raizo62, may I ask you to test again the latest usbtest branch? I reworked how interactiveness is detected. Needs extensive testing with telnet/ssh also.

**Comment by Raizo62** [comment]
re

With telnet, it seems to work (but the first fastethernet0/0 is always uncolored)

It is more complicated for me to test ssh

For usb : 
- when i use "tab", the line of command becomes uncolored.
- all config displayed by "show run" is not colored : 
![2020-03-23-112021_1920x1080_scrot](https://user-images.githubusercontent.com/7449584/77307481-9d4dce80-6cf9-11ea-835f-e398a1e3dce9.png)

[log.txt](https://github.com/realvitya/clicol/files/4368653/log.txt)

**Comment by realvitya** [comment]
Hm, in the provided log.txt everything looked colored.
When you produced this output on the picture: Did you use ENTER or SPACE to see more lines? I commanded the logic to buffer only if user hit ENTER. Not when he/she uses SPACE or TAB.
Please answer and I'll try to fix that!

**Comment by Raizo62** [comment]
to generate the log.txt, i use the clicol command finished by "&> log.txt", then i do not see my command and the output.
I suppose that i have push on enter several time to have all the "show run", before to close the connection

When i do "show run", cisco displays a colored "--More-" (1 -), and after a little second, displays le last "-" without color. If i push on Enter, the next lines are colored until next "--More-"
If i don't wait the last "-", and push on Enter, the "interface" is not colored but "fastethernet0/0" is colored

"show run" where i put only on Enter : 
[logenter.txt](https://github.com/realvitya/clicol/files/4368942/logenter.txt)
"show run" where i put only on Space : 
[logspace.txt](https://github.com/realvitya/clicol/files/4368943/logspace.txt)

**Comment by realvitya** [comment]
@Raizo62 would you please checkout latest usbtest and check again? I could reproduce only char by char input in GNS3 console connection. But now its ok for me. With space or enter either.
If your serial is slow, try to increase maxwait to default first (1.0) or higher (2.0 should be more than enough) These higher values are good for really slow network connections like VSAT or high packet drops.
Basically maxwait should be small only for local terminal. Like you want to run bash or other faster local program.

**Comment by Raizo62** [comment]
It is better.
I have tested with maxwait=1.0 and 2.0
When i do a "show run" on a switch with usb

- If i use "space"  to bypass the "--more-", the colors seem good
- If i use "enter" to bypass, the word "interface" is uncolored, and the "fastethernet0/0" is colored.

When i use "tab" to do the completion, all the prompt line becomes uncolored.

Do you need debug log or screenshot ?

**Comment by realvitya** [comment]
Yes please! Can you do the logging like you turn on debug and do it
interactively to show the problem. Then copy paste the screen. I know its
inconvenient, but helps me to understand what's happening.
Both for entered pager and an example for the TAB!
Thanks!

Raizo62 <notifications@github.com> ezt írta (időpont: 2020. márc. 26., Cs,
12:14):

**Comment by Raizo62** [comment]
> I know its inconvenient, but helps me to understand what's happening.

No problemo. Don't hesitate to ask what you need

![2020-03-27-101815_1366x768_scrot](https://user-images.githubusercontent.com/7449584/77741628-7fdc7580-7015-11ea-8272-d5513f9c6912.png)

logs got with copy paste : [copy_paste.txt](https://github.com/realvitya/clicol/files/4392556/copy_paste.txt)

logs got with the linux command tee : [tee.log](https://github.com/realvitya/clicol/files/4392570/tee.log)

**Comment by realvitya** [comment]
Sorry, I couldn't find in the debugs the points when you hit tab. At least
I don't see TAB "\t" character in your debugs.
So to summarize we have these two issues:

   1. If i use "enter" to bypass, the word "interface" is uncolored, and
   the "fastethernet0/0" is colored.
   This seems more tricky. I couldn't reproduce. I need another output
   showing this. please set debug to 2!
   2. When i use "tab" to do the completion, all the prompt line becomes
   uncolored.
   This can be because when you hit TAB, then device send a bunch of
   backspaces, then write the whole line again. Because in your case it is
   done char by char, we can't realize if we have a prompt already. What I can
   do is a workaround to check buffer for a prompt. Let me work on that.


Raizo62 <notifications@github.com> ezt írta (időpont: 2020. márc. 27., P,
10:30):

**Comment by realvitya** [comment]
@Raizo62 , please checkout again and test!
I did this: reworked some part around prompt and pager handling. Also I decided to leave pasting uncolored by intention. So nothing can interfere with pasting configuration.
For me, all glitches went away. I'm curious what you see.

**Comment by Raizo62** [comment]
I have tested.
The tail of the screenshot : 
The coloration of prompt works when i use tab.
When i do "show runn", "interface fast0/0" is colored when i use enter or space.
It is ok.

The top of the screenshot : 
"But", in the "show run", when i stays on enter keyboard, the "show run" is colored correctly, but after, many "prompt" is uncolored
![2020-03-29-104051_1920x1080_scrot](https://user-images.githubusercontent.com/7449584/77845878-67ee2880-71b2-11ea-98d8-7b5edacc9084.png)

**Comment by realvitya** [comment]
Would you please try again? I tried to do cleanup and addressing your
problem. Try moreover special characters like CTRL-A,U,L,D,E
Thanks so much!

Raizo62 <notifications@github.com> ezt írta (időpont: 2020. márc. 29., V,
11:46):

**Comment by realvitya** [comment]
Would you mind reaching out on [Discord](https://discord.gg/FbUNfkb)?

**Comment by Raizo62** [comment]
I have tested : it seems ok for all. Thanks :-D

> Would you mind reaching out on Discord?

To write english is not easy for me. My oral english is very bad. I prefer "text" chats

**Comment by Raizo62** [comment]
> Try moreover special characters like CTRL-A,U,L,D,E

I have not tested that, yet

**Comment by Raizo62** [comment]
CTRL-A : begin of line : ok (the prompt stays colored)
CTRL+U : delete line : ok
CTRL+L : Reprint the line : ok
CTRL-D : delete character : ok
CTRL-E : end of line : ok

**Comment by realvitya** [comment]
Can I close this bug item? What do you think? I'll test this branch in our
production and if all goes well, I'll merge it to master branch and release
a new bugfix version...

Raizo62 <notifications@github.com> ezt írta (időpont: 2020. márc. 30., H,
11:24):

**Comment by Raizo62** [comment]
> Can I close this bug item?

Yes, i think that is good. Thank you :-D

> I'll test this branch in our production and if all goes well, I'll merge it to master branch and release a new bugfix version...

Ok.

Thanks again
